### PR TITLE
fix: close the watermark fetcher and publishers after all the forwarders exit

### LINKED ISSUE #848: Flaky unit test TestReduceDataForward_NonKeyed
**Describe the bug**
See https://github.com/numaproj/numaflow/actions/runs/5535477672/attempts/2?pr=847

```
=== RUN   TestReduceDataForward_NonKeyed
{"level":"info","ts":"2023-07-12T19:32:30.310450855Z","logger":"numaflow","caller":"reduce/data_forward.go:567","msg":"Stopping reduce data forwarder..."}
{"level":"info","ts":"2023-07-12T19:32:30.310528355Z","logger":"numaflow","caller":"reduce/data_forward.go:572","msg":"Closed buffer reader","bufferFrom":"source-reduce-buffer"}
{"level":"info","ts":"2023-07-12T19:32:30.310589755Z","logger":"numaflow","caller":"fetch/edge_fetcher.go:218","msg":"Closing edge watermark fetcher","bucketName":"source-reduce-buffer"}
{"level":"info","ts":"2023-07-12T19:32:30.310635755Z","logger":"numaflow","caller":"publish/publisher.go:281","msg":"Closing watermark publisher","entityID":"reduce-to-vertex","otStore":"reduce-to-vertex_OT","hbStore":"reduce-to-vertex_PROCESSORS"}
{"level":"error","ts":"2023-07-12T19:32:30.310683056Z","logger":"numaflow","caller":"publish/publisher.go:296","msg":"Failed to delete the key in the heartbeat bucket","entityID":"reduce-to-vertex","otStore":"reduce-to-vertex_OT","hbStore":"reduce-to-vertex_PROCESSORS","bucket":"reduce-to-vertex_PROCESSORS","key":"reduce-to-vertex","error":"key reduce-to-vertex not found","stacktrace":"github.com/numaproj/numaflow/pkg/watermark/publish.(*publish).Close\n\t/home/runner/work/numaflow/numaflow/pkg/watermark/publish/publisher.go:296\ngithub.com/numaproj/numaflow/pkg/reduce.(*DataForward).ShutDown\n\t/home/runner/work/numaflow/numaflow/pkg/reduce/data_forward.go:585\ngithub.com/numaproj/numaflow/pkg/reduce.(*DataForward).Start\n\t/home/runner/work/numaflow/numaflow/pkg/reduce/data_forward.go:133"}
{"level":"error","ts":"2023-07-12T19:32:30.310776656Z","logger":"numaflow","caller":"reduce/data_forward.go:586","msg":"Failed to close watermark publisher","error":"key reduce-to-vertex not found","stacktrace":"github.com/numaproj/numaflow/pkg/reduce.(*DataForward).ShutDown\n\t/home/runner/work/numaflow/numaflow/pkg/reduce/data_forward.go:586\ngithub.com/numaproj/numaflow/pkg/reduce.(*DataForward).Start\n\t/home/runner/work/numaflow/numaflow/pkg/reduce/data_forward.go:133"}
panic: send on closed channel

goroutine 1431 [running]:
github.com/numaproj/numaflow/pkg/watermark/store/inmem.(*inMemStore).PutKV(0xc0036fca00, {0x3?, 0x3?}, {0x168de28, 0x10}, {0xc00016f300, 0x15, 0x0?})
	/home/runner/work/numaflow/numaflow/pkg/watermark/store/inmem/kv_store.go:134 +0x230
github.com/numaproj/numaflow/pkg/watermark/publish.(*publish).PublishWatermark(0xc0036ef710, {0xc0037146f0?, 0x168de28?, 0x0?}, {0x18d2608, 0xc000f45e70}, 0x14f2070?)
	/home/runner/work/numaflow/numaflow/pkg/watermark/publish/publisher.go:155 +0x9dd
github.com/numaproj/numaflow/pkg/reduce/pnf.(*processAndForward).publishWM(0xc0008dadd0, {0x18d1f08, 0xc0036c8300}, {0x10?, 0x1?, 0x0?}, 0x1?)
	/home/runner/work/numaflow/numaflow/pkg/reduce/pnf/processandforward.go:294 +0x6c3
github.com/numaproj/numaflow/pkg/reduce/pnf.(*processAndForward).Forward(0xc0008dadd0, {0x18d1f08?, 0xc0036c8300})
	/home/runner/work/numaflow/numaflow/pkg/reduce/pnf/processandforward.go:161 +0x6e5
github.com/numaproj/numaflow/pkg/reduce/pnf.(*OrderedProcessor).forward(0xc0036fe820, {0x18d1f08, 0xc0036c8300?})
	/home/runner/work/numaflow/numaflow/pkg/reduce/pnf/ordered.go:206 +0x148e
created by github.com/numaproj/numaflow/pkg/reduce/pnf.NewOrderedProcessor
	/home/runner/work/numaflow/numaflow/pkg/reduce/pnf/ordered.go:93 +0x65c
FAIL	github.com/numaproj/numaflow/pkg/reduce	15.905s
?   	github.com/numaproj/numaflow/pkg/reduce/applier	[no test files]
```

### LINKED ISSUE #855: coarse lock held during pbqManager deregister
The deregister function in `pbqmanager.go` holds the lock over filesystem calls which is quite detrimental.

### LINKED ISSUE #918: flaky test: Test_Single
https://github.com/numaproj/numaflow/actions/runs/5755632151/job/15603481806

```
=== RUN   Test_Single
{"level":"info","ts":"2023-08-03T21:08:35.559881977Z","logger":"numaflow","caller":"publish/publisher.go:71","msg":"Creating a new watermark publisher","entityID":"-0","otStore":"noop","hbStore":"noop"}
{"level":"info","ts":"2023-08-03T21:08:35.56094396Z","logger":"numaflow","caller":"nats/nats.go:158","msg":"Connecting to nats service..."}
{"level":"info","ts":"2023-08-03T21:08:35.561259084Z","logger":"numaflow","caller":"publish/publisher.go:265","msg":"Publishing Heartbeat ticker started","entityID":"-0","otStore":"noop","hbStore":"noop"}
    nats_test.go:115: timeout waiting for messages
{"level":"info","ts":"2023-08-03T21:08:45.670822271Z","logger":"numaflow","caller":"nats/nats.go:264","msg":"Shutting down nats source server..."}
{"level":"info","ts":"2023-08-03T21:08:45.670914978Z","logger":"numaflow","caller":"publish/publisher.go:281","msg":"Closing watermark publisher","entityID":"-0","otStore":"noop","hbStore":"noop"}
{"level":"info","ts":"2023-08-03T21:08:45.671078391Z","logger":"numaflow","caller":"nats/nats.go:273","msg":"Nats source server shutdown"}
{"level":"info","ts":"2023-08-03T21:08:45.671[5435](https://github.com/numaproj/numaflow/actions/runs/5755632151/job/15603481806#step:7:5436)28Z","logger":"numaflow","caller":"nats/nats.go:113","msg":"Nats disconnected"}
--- FAIL: Test_Single (10.14s)
```

### PR DESCRIPTION
**Pull request by yhl25**
- remove watcher dependency from fetcher since it's only used during close and move the close logic to processor manager stop
- introduced doneCh channel to gracefully stop all the watchers
- introduced doneCh and waitCh channels to gracefully stop the processor manager
- close the fetcher and publisher only after all the forward go routines exit
- renamed `WatermarkKVEntry`, `WatermarkKVStorer` and `WatermarkKVWatcher` to `KVEntry`, `KVStorer` and `KVWatcher`  and moved them out of watermark repo and introduced a separate package `kvs`, since they will be used for side input 
- update inmem KV store close logic to avoid panic 
- fixed a few flaky unit tests
- remove coarse lock inside pbqManager `Deregister()`

fixes #918 
fixes #848 
fixes #855

### CODE REVIEW

File: pkg/shared/kvs/interface.go [hunk 1]
```diff
@@ -0,0 +1,81 @@
...
+*/
+
+package kvs
+
+import (
+	"context"
+)
+
+// KVStore defines the storage for publishing the watermark and sideinput
+type KVStore interface {
```
Comment (vigith) [inline pkg/shared/kvs/interface.go:24]:
```suggestion
type KVStorer interface {
```

**Comment by jy4096** [comment]
could you add a brief summary in the PR description?

File: pkg/watermark/fetch/edge_fetcher.go [hunk 2]
```diff
@@ -63,11 +60,10 @@ func NewEdgeFetcher(ctx context.Context, bucketName string, storeWatcher store.W
 
 	return &edgeFetcher{
 		ctx:              ctx,
-		bucketName:       bucketName,
-		storeWatcher:     storeWatcher,
+		bucketName:       manager.GetBucket(),
 		processorManager: manager,
 		lastProcessedWm:  lastProcessedWm,
-		log:              log.With("bucket", bucketName),
+		log:              log.With("bucket", manager.GetBucket()),
```
Comment (jy4096) [inline pkg/watermark/fetch/edge_fetcher.go:66]: Do we still need this if we have `log := logging.FromContext(ctx).With("bucketName", manager.GetBucket())` ??....

Comment (yhl25) [reply pkg/watermark/fetch/edge_fetcher.go]: removed

File: pkg/reduce/pnf/processandforward_test.go [hunk 3]
```diff
@@ -448,11 +449,11 @@ func createProcessAndForwardAndOTStore(ctx context.Context, key string, pbqManag
 }
 
 // buildPublisherMap builds OTStore and publisher for each toBuffer
```
Comment (jy4096) [inline pkg/reduce/pnf/processandforward_test.go:451]:
```suggestion
// buildPublisherMapAndOTStore builds OTStore and publisher for each toBuffer
```

File: pkg/shared/kvs/jetstream/kv_store.go [hunk 4]
```diff
@@ -24,14 +24,14 @@ import (
...
 
 	jsclient "github.com/numaproj/numaflow/pkg/shared/clients/nats"
 	"github.com/numaproj/numaflow/pkg/shared/logging"
-	"github.com/numaproj/numaflow/pkg/watermark/store"
 )
 
-// jetStreamStore implements the watermark's KV store backed up by Jetstream.
+// jetStreamStore implements the KV store backed up by Jetstream.
 type jetStreamStore struct {
 	pipelineName string
```
Comment (jy4096) [inline pkg/shared/kvs/jetstream/kv_store.go:36]: looks like we don't use this `pipelineName` anywhere else?

File: pkg/shared/kvs/jetstream/kv_watch.go [hunk 5]
```diff
@@ -175,6 +179,10 @@ func (jsw *jetStreamWatch) Watch(ctx context.Context) (<-chan store.WatermarkKVE
 				// reset the timer, since we have drained the timer channel its safe to reset it
 				jsw.kvwTimer.Reset(jsw.opts.watcherCreationThreshold)
 
+			case <-jsw.doneCh:
+				jsw.log.Infow("stopping WatchAll", zap.String("watcher", jsw.GetKVName()))
```
Comment (whynowy) [inline pkg/shared/kvs/jetstream/kv_watch.go:183]: Capital the first letter for logging.

**Review by vigith — APPROVED** [review]

**Pull request merged**